keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+    DriverName string
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource that uses this class. The driver will
+    // be able to distinguish between parameters stored here and
+    // and those stored in ResourceClaimSpec. These parameters
+    // here can only be set by cluster administrators.
+    Parameters runtime.RawExtension
+}
+
+type ResourceClaim struct {
+    // The driver must set a finalizer here before it attempts to
+    // allocate the resource. It removes the finalizer again when
+    // a) the allocation attempt has definitely failed or b) when
+    // the allocated resource was freed. This ensures that
+    // resources are not leaked.
+    ObjectMeta
+    // Spec describes the desired attributes of a resource that then
+    // needs to be allocated. It can only be set once when creating
+    // the ResourceClaim.
+    Spec ResourceClaimSpec
+    // Status describes whether the resource is available and with
+    // which attributes.
+    Status ResourceClaimStatus
+}
+
+type ResourceClaimSpec struct {
+    // ResourceClassName references the driver and additional
+    // parameters via the name of a ResourceClass that was
+    // created as part of the driver deployment.
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a driver-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
+
+    // Allocation can start immediately or when a Pod wants to use
+    // the resource. Waiting for a Pod is the default.
+    AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated
+// immediately when it gets created (AllocationModeImmediate)
+// or whether allocation is delayed until it is needed for
+// a Pod (AllocationModeDelayed). Other modes might get
+// added in the future.
+type AllocationMode string
+
+const (
+     // When a ResourceClass has AllocationModeImmediate,
+     // allocation of ResourceClaims with that ResourceClass
+     // starts as soon as the ResourceClaim gets created.
+     // This is done without considering the needs of
+     // Pods that will use the ResourceClaim because those
+     // Pods are not known yet.
+     AllocationModeImmediate AllocationMode = “Immediate”
+
+     // When a ResourceClass has AllocationModeDelayed,
+     // allocation of ResourceClaims with that ResourceClass
+     // is delayed until a Pod gets scheduled that needs the
+     // ResourceClaim. The scheduler will consider all
+     // resource requirements of that Pod and trigger
+     // allocation for a node that fits the Pod.
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
```

Comment:
which actor writes status?  kubelet, scheduler, driver, all of the above?

Reply:
All of them, in different ways. Some fields are written by just one actor and read by the others, other fields are maintained together.

Reply:
I split up the API into different structs to make this clearer.